tests/unittests/GraphOptzTest.cpp

```diff
@@ -373,17 +373,21 @@ TEST_F(GraphOptz, optimizeBatchNormAfterConvWithPred) {
   EXPECT_EQ(F_->getNodes().size(), 3);
 
   ::glow::convertPlaceholdersToConstants(F_, bindings_, {});
-  ::glow::optimize(F_, CompilationMode::Infer);
-  EXPECT_EQ(F_->getNodes().size(), 2);
+  optimizedF_ = optimizeFunction(F_);
```

Comment:
I think you *might* have to clone the function manually and then call `convertPlaceholdersToConstants` and `optimize` on the clone. Based on a cursory examination of the optimization I think it should be fine, but it's a good idea to check the dot graphs of `F_` and `optimizedF_` to confirm.

Reply:
For some reason I am not able to upload the file. May be due to company firewall. Pasting them here.

F_.dot:
digraph DAG {
	rankdir=TB;
conv[
	label = "{{<Input>Input|<Filter>Filter|<Bias>Bias}|{Convolution\lname : conv\lPredicate : Yes\lInput : float\<1 x 10 x 20 x 3\>\lFilter : float\<16 x 5 x 5 x 3\>\lBias : float\<16\>\lKernels : [5, 5]\lStrides : [1, 1]\lPads : [2, 2, 2, 2]\lGroup : 1\lDilation : 1\lLayout : NHWC\lFusedActivation : \lusers : 1\lResult : float\<1 x 10 x 20 x 16\>\l}|{<Result>Result}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Thistle1
penwidth = 2];

predicate[
	label = "{{Placeholder\lname : \"predicate\"\llayout : *\loutput : float\<1\>\lusers : 1\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

A[
	label = "{{Placeholder\lname : \"A\"\llayout : *\loutput : float\<1 x 10 x 20 x 3\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

filter__1[
	label = "{{Constant\lname : \"filter__1\"\llayout : *\loutput : float\<16 x 5 x 5 x 3\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

bias__1[
	label = "{{Constant\lname : \"bias__1\"\llayout : *\loutput : float\<16\>\lusers : 2\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

batch[
	label = "{{<Input>Input|<Scale>Scale|<Bias>Bias|<Mean>Mean|<Var>Var}|{BatchNormalization\lname : batch\lPredicate : Yes\lInput : float\<1 x 10 x 20 x 16\>\lScale : float\<16\>\lBias : float\<16\>\lMean : float\<16\>\lVar : float\<16\>\lChannelIdx : 3\lEpsilon : 1.000000e-04\lMomentum : 9.000000e-01\lusers : 1\lResult : float\<1 x 10 x 20 x 16\>\l}|{<Result>Result}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=PaleTurquoise2
penwidth = 2];

predicate__1[
	label = "{{Placeholder\lname : \"predicate__1\"\llayout : *\loutput : float\<1\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

scale__1[
	label = "{{Constant\lname : \"scale__1\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

mean__1[
	label = "{{Constant\lname : \"mean__1\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

variance__1[
	label = "{{Constant\lname : \"variance__1\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

ret_save[
	label = "{{<Input>Input|<Output>Output}|{Save\lname : ret_save\lInput : float\<1 x 10 x 20 x 16\>\lOutput : float\<1 x 10 x 20 x 16\>\lusers : 0\l}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=PowderBlue
penwidth = 2];

ret[
	label = "{{Placeholder\lname : \"ret\"\llayout : *\loutput : float\<1 x 10 x 20 x 16\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

ret:Output -> ret_save:Output [dir="both"];
variance__1:Output -> batch:Var;
filter__1:Output -> conv:Filter;
mean__1:Output -> batch:Mean;
A:Output -> conv:Input;
scale__1:Output -> batch:Scale;
bias__1:Output -> batch:Bias;
conv:Result -> batch:Input;
predicate__1:Output -> batch:w;
batch:Result -> ret_save:Input;
predicate:Output -> conv:w;
bias__1:Output -> conv:Bias;
}

optimizedF_.dot:
digraph DAG {
	rankdir=TB;
conv[
	label = "{{<Input>Input|<Filter>Filter|<Bias>Bias}|{Convolution\lname : conv\lPredicate : Yes\lInput : float\<1 x 10 x 20 x 3\>\lFilter : float\<16 x 5 x 5 x 3\>\lBias : float\<16\>\lKernels : [5, 5]\lStrides : [1, 1]\lPads : [2, 2, 2, 2]\lGroup : 1\lDilation : 1\lLayout : NHWC\lFusedActivation : \lusers : 1\lResult : float\<1 x 10 x 20 x 16\>\l}|{<Result>Result}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Thistle1
penwidth = 2];

predicate__1[
	label = "{{Placeholder\lname : \"predicate__1\"\llayout : *\loutput : float\<1\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

A[
	label = "{{Placeholder\lname : \"A\"\llayout : *\loutput : float\<1 x 10 x 20 x 3\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

filter__2[
	label = "{{Constant\lname : \"filter__2\"\llayout : *\loutput : float\<16 x 5 x 5 x 3\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

bias__2[
	label = "{{Constant\lname : \"bias__2\"\llayout : *\loutput : float\<16\>\lusers : 1\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=Snow3 color=DeepSkyBlue4
penwidth = 2];

ret_save[
	label = "{{<Input>Input|<Output>Output}|{Save\lname : ret_save\lInput : float\<1 x 10 x 20 x 16\>\lOutput : float\<1 x 10 x 20 x 16\>\lusers : 0\l}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=PowderBlue
penwidth = 2];

ret[
	label = "{{Placeholder\lname : \"ret\"\llayout : *\loutput : float\<1 x 10 x 20 x 16\>\lusers : 2\ltrainable : 0\l}|{<Output>Output}}"
	shape = "record"
	style="filled,rounded"
	fillcolor=MistyRose1
penwidth = 2];

ret:Output -> ret_save:Output [dir="both"];
conv:Result -> ret_save:Input;
filter__2:Output -> conv:Filter;
bias__2:Output -> conv:Bias;
A:Output -> conv:Input;
predicate__1:Output -> conv:w;
}

Reply:
Uh, can you grab a screenshot and post that? I think you can drag and drop the image into the text box.

Reply:
<img width="936" alt="optimizedF_" src="https://user-images.githubusercontent.com/60203774/73393183-1ac40800-4290-11ea-90c6-74b5978dc84b.png">
<img width="927" alt="F_" src="https://user-images.githubusercontent.com/60203774/73393198-20215280-4290-11ea-951a-8403e155f56d.png">